### PR TITLE
IPA build workflow

### PR DESCRIPTION
**Pull request by assumptionsandg**

### CODE REVIEW

File: .github/workflows/ipa-image-build.yml [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        run: |
+          mkdir -p venvs &&
+          pushd venvs &&
+          python3 -m venv kayobe &&
+          source kayobe/bin/activate &&
+          pip install -U pip &&
+          pip install -r ../src/kayobe-config/requirements.txt
+
+      - name: Install terraform
+        uses: hashicorp/setup-terraform@v2
```
Comment (Alex-Welsh) [inline .github/workflows/ipa-image-build.yml]:
```suggestion
        uses: hashicorp/setup-terraform@v3
```

File: .github/workflows/ipa-image-build.yml [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+        if: steps.build_ubuntu_jammy_ipa.outcome == 'failure'
+
+      - name: Upload Ubuntu 22.04 Jammy IPA kernel image to Ark
+        run: |
+          source venvs/kayobe/bin/activate &&
+          source src/kayobe-config/kayobe-env --environment ci-builder &&
+          kayobe playbook run \
+          src/kayobe-config/etc/kayobe/ansible/pulp-artifact-upload.yml \
+          -e artifact_path=/opt/kayobe/images/ipa \
+          -e artifact_type=ipa \
```
Comment (Alex-Welsh) [inline .github/workflows/ipa-image-build.yml]: Just a suggestion to match the existing formatting of `kayobe-images`
```suggestion
          -e artifact_type=ipa-images \
```

File: .github/workflows/ipa-image-build.yml [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        if: inputs.ubuntu-jammy && steps.build_ubuntu_jammy_ipa.outcome == 'success'
+
+      - name: Upload Ubuntu 22.04 Jammy IPA ramdisk image to Ark
+        run: |
+          source venvs/kayobe/bin/activate &&
+          source src/kayobe-config/kayobe-env --environment ci-builder &&
+          kayobe playbook run \
+          src/kayobe-config/etc/kayobe/ansible/pulp-artifact-upload.yml \
+          -e artifact_path=/opt/kayobe/images/ipa \
+          -e artifact_type=ipa \
```
Comment (Alex-Welsh) [inline .github/workflows/ipa-image-build.yml]: ditto

File: doc/source/configuration/ipa.rst [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+=========================
+Ironic Python Agent (IPA)
+=========================
+
+Release Train IPA
+=================
+
+StackHPC provides built Ironic Python Agent (IPA) images in Release Train
+through Ark.
```
Comment (Alex-Welsh) [inline doc/source/configuration/ipa.rst]:
```suggestion
StackHPC provides prebuilt Ironic Python Agent (IPA) images in Release Train
through Ark.
```

File: doc/source/configuration/ipa.rst [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+These images are built in a CI using a Kayobe workflow :kayobe-doc:
+`Kayobe documentation <configuration/reference/ironic-python-agent.html>`
+and are configured in this repository.
```
Comment (Alex-Welsh) [inline doc/source/configuration/ipa.rst]:
```suggestion
These images are built in CI using a GitHub workflow and are configured in this
repository. See  :kayobe-doc: `Kayobe documentation
<configuration/reference/ironic-python-agent.html>` for more details on IPA.
```

File: doc/source/configuration/ipa.rst [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+Release Train IPA is used by Bifrost and Overcloud Ironic by default in
+StackHPC Kayobe Configuration, and is pulled in to Inspector when running
+``kayobe seed service deploy`` for Bifrost or ``kayobe overcloud post configure``
+for Ironic. This behaviour can be disabled in `stackhpc-ipa-images.yml`:
```
Comment (Alex-Welsh) [inline doc/source/configuration/ipa.rst]:
```suggestion
Release Train IPA images are used by Bifrost and Overcloud Ironic by default in
the StackHPC Kayobe Configuration, and are pulled in to the Ironic Inspector
when running ``kayobe seed service deploy`` for Bifrost or ``kayobe overcloud
post configure`` for Overcloud Ironic. This behaviour can be disabled in
`stackhpc-ipa-images.yml`:
```

File: etc/kayobe/ansible/pulp-artifact-upload.yml [hunk 7]
```diff
@@ -36,6 +37,31 @@
...
+      ansible.builtin.copy:
+        content: "{{ file_stats.stat.checksum }}"
+        dest: "/tmp/{{ found_files.files[0].path | basename }}.sha256"
+      when: upload_checksum
+
+    - name: Get sha256 hash for checksum
+      ansible.builtin.stat:
+        path: "/tmp/{{ found_files.files[0].path | basename }}.sha256"
+        checksum_algorithm: sha256
+      register: checksum_stats
```
Comment (Alex-Welsh) [inline etc/kayobe/ansible/pulp-artifact-upload.yml:50]: 0_o

File: etc/kayobe/bifrost.yml [hunk 8]
```diff
@@ -135,7 +135,7 @@
 # Ironic Python Agent (IPA) configuration.
 
 # URL of Ironic Python Agent (IPA) kernel image.
-#kolla_bifrost_ipa_kernel_upstream_url:
+kolla_bifrost_ipa_kernel_upstream_url: "{% if stackhpc_ipa_image_bifrost_enabled | bool %}{{ stackhpc_ipa_image_url }}/ipa.kernel{% endif %}"
```
Comment (Alex-Welsh) [inline etc/kayobe/bifrost.yml]: I think we use this style of conditional more in SKC
```suggestion
kolla_bifrost_ipa_kernel_upstream_url: "{{ (stackhpc_ipa_image_url + '/ipa.kernel') if stackhpc_ipa_image_bifrost_enabled | bool }}"
```

File: etc/kayobe/bifrost.yml [hunk 9]
```diff
@@ -144,7 +144,7 @@
 #kolla_bifrost_ipa_kernel_checksum_algorithm:
 
 # URL of Ironic Python Agent (IPA) ramdisk image.
-#kolla_bifrost_ipa_ramdisk_upstream_url:
+kolla_bifrost_ipa_ramdisk_upstream_url: "{% if stackhpc_ipa_image_bifrost_enabled | bool %}{{ stackhpc_ipa_image_url }}/ipa.initramfs{% endif %}"
```
Comment (Alex-Welsh) [inline etc/kayobe/bifrost.yml]:
```suggestion
kolla_bifrost_ipa_ramdisk_upstream_url: "{{ (stackhpc_ipa_image_url + '/ipa.initramfs') if stackhpc_ipa_image_bifrost_enabled | bool }}"
```

File: etc/kayobe/environments/ci-builder/inventory/hosts [hunk 10]
```diff
@@ -3,5 +3,9 @@
 [ofed-builder:children]
 seed
 
+# This is neccesary to run `overcloud deployment image build`
```
Comment (Alex-Welsh) [inline etc/kayobe/environments/ci-builder/inventory/hosts:6]: have you checked this won't break any other workflows?

File: etc/kayobe/ipa.yml [hunk 11]
```diff
@@ -87,7 +87,7 @@ ipa_build_dib_elements_extra:
 #ipa_images_kernel_name:
 
 # URL of Ironic deployment kernel image to download.
-#ipa_kernel_upstream_url:
+ipa_kernel_upstream_url: "{% if stackhpc_ipa_image_overcloud_enabled | bool %}{{ stackhpc_ipa_image_url }}/ipa.kernel{% endif %}"
```
Comment (Alex-Welsh) [inline etc/kayobe/ipa.yml]: See previous comments

File: etc/kayobe/ipa.yml [hunk 12]
```diff
@@ -99,7 +99,7 @@ ipa_build_dib_elements_extra:
 #ipa_images_ramdisk_name:
 
 # URL of Ironic deployment ramdisk image to download.
-#ipa_ramdisk_upstream_url:
+ipa_ramdisk_upstream_url: "{% if stackhpc_ipa_image_overcloud_enabled | bool %}{{ stackhpc_ipa_image_url }}/ipa.initramfs{% endif %}"
```
Comment (Alex-Welsh) [inline etc/kayobe/ipa.yml]: See previous comments

File: etc/kayobe/stackhpc-ipa-images.yml [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+# Whether to use Release Train IPA images for Bifrost
+stackhpc_ipa_image_bifrost_enabled: true
+
+# Whether to use Release Train IPA images for Overcloud Ironic
+stackhpc_ipa_image_overcloud_enabled: true
+
+# The IPA image source, defined by os_distribution,
+# os_release, and the current stable version.
+stackhpc_ipa_image_url: "{{ stackhpc_release_pulp_content_url }}/ipa/\
```
Comment (Alex-Welsh) [inline etc/kayobe/stackhpc-ipa-images.yml]: If you change the artifact type:
```suggestion
stackhpc_ipa_image_url: "{{ stackhpc_release_pulp_content_url }}/ipa-images/\
```

File: releasenotes/notes/ipa-build-workflow-cde24f3b9e3e0667.yaml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
```
Comment (Alex-Welsh) [inline releasenotes/notes/ipa-build-workflow-cde24f3b9e3e0667.yaml:3]: Needs a second list item saying that the stackhpc ipa images are enabled by default

**Review by Alex-Welsh — CHANGES_REQUESTED** [review]
Great work on this. Mostly just styling and nit comments

File: .github/workflows/ipa-image-build.yml (hunk 2 above)
Comment (holliefae) [reply .github/workflows/ipa-image-build.yml]: We should probably do this for Amphora too, for future reference. @seunghun1ee

File: etc/kayobe/ansible/pulp-artifact-upload.yml (hunk 7 above)
Comment (holliefae) [reply etc/kayobe/ansible/pulp-artifact-upload.yml:50]: Unfortunately this seems to be required in squeezer: https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/12201646661/job/34040505888#step:27:65

This is bad though.

File: etc/kayobe/environments/ci-builder/inventory/hosts (hunk 10 above)
Comment (holliefae) [reply etc/kayobe/environments/ci-builder/inventory/hosts:6]: I'll run the host image/container build of this branch and see how it goes.

File: etc/kayobe/ansible/pulp-artifact-upload.yml (hunk 7 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/pulp-artifact-upload.yml:50]: I can't see a better way of doing it. It ain't pretty, but I think the logic of it

File: etc/kayobe/environments/ci-builder/inventory/hosts (hunk 10 above)
Comment (holliefae) [reply etc/kayobe/environments/ci-builder/inventory/hosts:6]: Turns out this results in Ark repos being enabled due to being in `overcloud` now. This is now disabled [b0cf130](https://github.com/stackhpc/stackhpc-kayobe-config/pull/1414/commits/b0cf130d15f8c1d187a3ea7b90909243e468fdef)

Host image build here https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/12315268297/job/34373114113

**Review by Alex-Welsh — APPROVED** [review]
Excellent work Jake

**Pull request merged**